### PR TITLE
Add '-Fsystem' framework search option to indicate path for frameworks that should be treated as 'system'

### PR DESCRIPTION
**Pull request by akyrtzi**
This has the effect of propagating the search path to the clang importer as '-iframework'.
It doesn't affect whether a swift module is treated as system or not, this can be done as follow-up enhancement.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci smoke test

**Comment by akyrtzi** [comment]
@jrose-apple could you review ?

File: include/swift/Option/Options.td [hunk 1]
```diff
@@ -148,6 +148,10 @@ def F : JoinedOrSeparate<["-"], "F">, Flags<[FrontendOption]>,
   HelpText<"Add directory to framework search path">;
 def F_EQ : Joined<["-"], "F=">, Flags<[FrontendOption]>, Alias<F>;
 
+def Fsystem : JoinedOrSeparate<["-"], "Fsystem">, Flags<[FrontendOption]>,
```
Comment (jrose-apple) [inline include/swift/Option/Options.td]: Nitpick: I think this one should probably never be joined.

File: lib/Driver/ToolChains.cpp [hunk 2]
```diff
@@ -930,6 +931,8 @@ toolchains::Darwin::constructInvocation(const InterpretJobAction &job,
                                      runtimeLibraryPath);
   addPathEnvironmentVariableIfNeeded(II.ExtraEnvironment, "DYLD_FRAMEWORK_PATH",
                                      ":", options::OPT_F, context.Args);
+  addPathEnvironmentVariableIfNeeded(II.ExtraEnvironment, "DYLD_FRAMEWORK_PATH",
+                                     ":", options::OPT_Fsystem, context.Args);
```
Comment (jrose-apple) [inline lib/Driver/ToolChains.cpp]: This won't combine them, unfortunately. I'm okay with this going unimplemented for now, though; please just file a JIRA bug about it.

File: lib/Driver/ToolChains.cpp [hunk 3]
```diff
@@ -124,6 +124,7 @@ static void addCommonFrontendArgs(const ToolChain &TC,
 
   inputArgs.AddAllArgs(arguments, options::OPT_I);
   inputArgs.AddAllArgs(arguments, options::OPT_F);
+  inputArgs.AddAllArgs(arguments, options::OPT_Fsystem);
```
Comment (jrose-apple) [inline lib/Driver/ToolChains.cpp]: The argument order matters here. Should we keep -F and -Fsystem options interleaved, or always put -Fsystem first or last?

File: lib/Frontend/CompilerInvocation.cpp [hunk 4]
```diff
@@ -1036,7 +1036,13 @@ static bool ParseSearchPathArgs(SearchPathOptions &Opts,
...
   for (const Arg *A : make_range(Args.filtered_begin(OPT_F),
                                  Args.filtered_end())) {
-    Opts.FrameworkSearchPaths.push_back(resolveSearchPath(A->getValue()));
+    Opts.FrameworkSearchPaths.push_back({resolveSearchPath(A->getValue()),
+                                         /*isSystem=*/false});
+  }
+  for (const Arg *A : make_range(Args.filtered_begin(OPT_Fsystem),
+                                 Args.filtered_end())) {
+    Opts.FrameworkSearchPaths.push_back({resolveSearchPath(A->getValue()),
+                                         /*isSystem=*/true});
```
Comment (jrose-apple) [inline lib/Frontend/CompilerInvocation.cpp]: Please do these in the same loop. (`filtered_begin` can take multiple arguments.)

File: tools/SourceKit/lib/SwiftLang/SwiftEditorInterfaceGen.cpp [hunk 5]
```diff
@@ -889,8 +889,12 @@ void SwiftLangSupport::findInterfaceDocument(StringRef ModuleName,
 
   const auto &SPOpts = Invocation.getSearchPathOptions();
   addArgPair("-sdk", SPOpts.SDKPath);
-  for (auto &Path : SPOpts.FrameworkSearchPaths)
-    addArgPair("-F", Path);
+  for (auto &FramePath : SPOpts.FrameworkSearchPaths) {
+    if (FramePath.IsSystem)
+      addArgPair("-iframework", FramePath.Path);
```
Comment (jrose-apple) [inline tools/SourceKit/lib/SwiftLang/SwiftEditorInterfaceGen.cpp]: This is a Swift argument list, right? :-)

File: include/swift/Option/Options.td (hunk 1 above)
Comment (akyrtzi) [inline include/swift/Option/Options.td]: Right, I found it surprising that -F= could be joined, I'd prefer if we avoided '=' joined aliases completely, but ship has sailed.

File: lib/Driver/ToolChains.cpp (hunk 3 above)
Comment (akyrtzi) [inline lib/Driver/ToolChains.cpp]: I think it's better to put -Fsystem last, so 'user' search paths get precedence.

File: include/swift/Option/Options.td (hunk 1 above)
Comment (jrose-apple) [inline include/swift/Option/Options.td]: This is about the *non* '=' version. You can say `-F/foo/bar/`, but I think `-Fsystem/foo/bar/` is probably a bad idea.

(I'm okay with not adding a `-Fsystem=` option too.)

File: lib/Driver/ToolChains.cpp (hunk 3 above)
Comment (jrose-apple) [inline lib/Driver/ToolChains.cpp]: Okay. That doesn't match Clang's behavior, but it is a simple story and a defensible position.

File: tools/SourceKit/lib/SwiftLang/SwiftEditorInterfaceGen.cpp (hunk 5 above)
Comment (akyrtzi) [inline tools/SourceKit/lib/SwiftLang/SwiftEditorInterfaceGen.cpp]: Oops, good catch!

File: include/swift/Option/Options.td (hunk 1 above)
Comment (akyrtzi) [inline include/swift/Option/Options.td]: Ah got it, good catch!

**Comment by akyrtzi** [comment]
@swift-ci smoke test

**Review by jrose-apple — APPROVED** [review]

**Comment by akyrtzi** [comment]
I have a commit to update lldb, I'll apply after merging this.

**Comment by akyrtzi** [comment]
Filed JIRA for the DYLD_FRAMEWORK_PATH issue: https://bugs.swift.org/browse/SR-3967

**Pull request merged**